[ONNX] Fix type annotations and enable type checking for all apis

justinchuby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84091

Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

Profile:

export `torchvision.models.alexnet(pretrained=True)`

```
with runtime type checking: 21.314 / 10 passes
without runtime type checking: 20.797 / 10 passes

+ 2.48%
```

justinchuby committed:
Annotate types

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Annotate types"




[ghstack-poisoned]

justinchuby commented on torch/onnx/symbolic_opset12.py:
```diff
@@ -90,17 +95,20 @@ def _dropout_returns_masked_input_and_mask(
-@symbolic_helper.parse_args("v", "f", "i")
+@symbolic_helper.parse_args("v", "f", "b")
+@_beartype.beartype
 def dropout(g, input, p, train):
```
Changed many "i" to "b" because (1) they are bool types (2) when translated to onnx Booleans are casted to int w/o issues.

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -115,9 +125,10 @@ def _ort_session(
-    ort_outs: Sequence[np.ndarray],
-    pt_outs: Sequence[torch.Tensor],
+    ort_outs: Union[_NumericType, Sequence[_NumericType], Sequence, Dict],
+    pt_outs: Union[_NumericType, Sequence[_NumericType], Sequence, Dict],
```
Relaxed types here

justinchuby replied:
Removed Dict

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -582,7 +612,9 @@ def verify(
     use_external_data: bool = False,
-    additional_test_inputs: Optional[Sequence[Tuple[Any, ...]]] = None,
+    additional_test_inputs: Optional[
+        Sequence[Union[torch.Tensor, Tuple[Any, ...]]]
```
Expanded

justinchuby commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -4823,11 +5096,12 @@ def linalg_norm(
     self: torch._C.Value,
     ord: float,
-    dim: List[int],
+    dim: Optional[List[int]],
```
Fixed

justinchuby commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -1432,15 +1523,17 @@ def symbolic_fn(g, input, kernel_size, stride, padding, dilation, ceil_mode):
-        padding: Union[int, Tuple[int, ...]],
+        kernel_size: Sequence[int],
+        stride: Sequence[int],
+        padding: Union[int, Sequence[int]],
```
Fixed

justinchuby commented on torch/onnx/symbolic_opset11.py:
```diff
@@ -487,29 +514,33 @@ def stack(g, tensor_list, dim):
-        padding: Union[int, Tuple[int, ...]],
+        kernel_size: Sequence[int],
+        stride: Sequence[int],
+        padding: Union[int, Sequence[int]],
```
Fixed

justinchuby commented on torch/onnx/symbolic_helper.py:
```diff
@@ -163,33 +165,41 @@ def _parse_arg(
     # NOTE: prim::Constant at this stage usually means something not compatible in ONNX,
     # otherwise it'd be converted to onnx::Constant
-    if _is_value(value) and _is_onnx_constant(value):
+    if isinstance(value, _C.Value) and _is_onnx_constant(value):
```
Changed for mypy checks on `_is_onnx_constant`

thiagocrepaldi replied:
At least add a TODO comment here and on `_is_value` describing the mypy issue for a future fix. Otherwise this will be forgotten

justinchuby replied:
Done

justinchuby commented on torch/onnx/symbolic_helper.py:
```diff
@@ -163,33 +165,41 @@ def _parse_arg(
-def _maybe_get_const(value: _C.Value, descriptor: _ValueDescriptor):
+@_beartype.beartype
+def _maybe_get_const(
+    value: Optional[Union[_C.Value, torch.Tensor, Number, Sequence]],
```
Relaxed

justinchuby replied:
I wanted to (1)enable typing checking before (2) fixing the code. Modifying the annotations is a way to minimal change code to enable checking without modifying current behaviors. We can then optimize stuff from there

justinchuby commented on torch/onnx/_patch_torch.py:
```diff
@@ -8,13 +8,15 @@
     g: torch._C.Graph,
     opname: str,
-    *raw_args: torch._C.Value,
+    *raw_args: Union[torch.Tensor, torch._C.Value],
```
Is this right? Can raw_args be torch.Tensor? (They exist in tests)

thiagocrepaldi replied:
I would say no, but could be wrong.
Which test has it? let's look into it

BowenBao replied:
Can we clean up and remove it (preferably in a separate PR)? I guess Meta internal test is likely to scream though :(

justinchuby replied:
Sure thing!

thiagocrepaldi replied:
Letting a non Value goes true the most generic function for symbolics open the door for many bugs. With this, any symbolic can wrongly pass `torch.Tensor` when it was not supposed to. IMO we need to fix this as opposed to relax the type checking to allow CI to pass. Hardening the code is always best, even if it needs more refactoring

justinchuby replied:
In a follow-up PR. This is already too big

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

TODO: profile performance

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

TODO: profile performance

[ghstack-poisoned]

justinchuby commented:
New

```
=================================== FAILURES ===================================
_ TestONNXRuntime_opset_version_11_is_script_True_keep_initializers_as_inputs_True.test_pad_types _
[gw1] linux -- Python 3.7.13 /opt/conda/bin/python
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 7409, in test_pad_types
    self.run_test(Pad(), (x, y))
  File "/var/lib/jenkins/workspace/test/onnx/onnx_test_common.py", line 132, in run_test
    _run_test(script_model, scripting_remained_onnx_input_idx, flatten=False)
  File "/var/lib/jenkins/workspace/test/onnx/onnx_test_common.py", line 116, in _run_test
    verbose=verbose,
  File "/var/lib/jenkins/workspace/test/onnx/onnx_test_common.py", line 44, in run_model_test
    return verification.verify(*args, **kwargs)
  File "<@beartype(torch.onnx.verification.verify) at 0x7f13a24f15f0>", line 495, in verify
  File "/opt/conda/lib/python3.7/site-packages/torch/onnx/verification.py", line 716, in verify
    acceptable_error_percentage,
  File "<@beartype(torch.onnx.verification._compare_ort_pytorch_model) at 0x7f13a24f0560>", line 33, in _compare_ort_pytorch_model
  File "/opt/conda/lib/python3.7/site-packages/torch/onnx/verification.py", line 337, in _compare_ort_pytorch_model
    compare_ort_pytorch_model_with_input(input_args, input_kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/onnx/verification.py", line 325, in compare_ort_pytorch_model_with_input
    ort_outs = _run_ort(ort_session, ort_inputs)
  File "/opt/conda/lib/python3.7/site-packages/torch/onnx/verification.py", line 100, in _run_ort
    ort_outs = ort_session.run(None, ort_inputs)
  File "/var/lib/jenkins/.local/lib/python3.7/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 200, in run
    return self._sess.run(output_names, input_feed, run_options)
RuntimeError: Iterable of <class 'numpy.ndarray'> should be given as array for input 'pad.1'.
```

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

TODO: profile performance

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

TODO: profile performance

[ghstack-poisoned]

thiagocrepaldi commented:
> Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
> 
> * **->** [[ONNX] Fix type annotations and enable type checking for all apis #84091](https://github.com/pytorch/pytorch/pull/84091)
> 
> Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.
> 
> TODO: profile performance

Add a comment in the places we cannot type check so that nobody tries or do it wrong

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

TODO: profile performance

[ghstack-poisoned]

justinchuby commented:
> Add a comment in the places we cannot type check so that nobody tries or do it wrong

Done

> Also adds a note stating it is mandatory to add type checking in all symbolics and their helpers are `symbolic_helper.py`, near the header doc
> 
> ```python
> # Note [Edit Symbolic Files]
> # EDITING THIS FILE AND SYMBOLIC_OPSET<VERSION> FILES? READ THIS FIRST!
> ```

Done

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

TODO: profile performance

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

TODO: profile performance

[ghstack-poisoned]

BowenBao reviewed (approved):
LGTM. I noticed that some new annotations were added, while others were not. Is that for passing certain check failures?

justinchuby commented:
> LGTM. I noticed that some new annotations were added, while others were not. Is that for passing certain check failures?

I may have missed some. Could you point them out?

justinchuby commented on torch/onnx/symbolic_opset11.py:
```diff
@@ -487,29 +514,33 @@ def stack(g, tensor_list, dim):
         padding = symbolic_helper._avgpool_helper(
             tuple_fn, padding, kernel_size, stride, divisor_override, name
         )
+        assert isinstance(padding, tuple)
```
That's ok imo. Because a tuple is a Sequence

justinchuby replied:
Unless tuple is treated differently?

BowenBao replied:
List is also Sequence, In my mind I actually thought it was List -> tuple, which is bit more concerning, but I think you are right.

justinchuby commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -2333,33 +2475,36 @@ def _layer_norm_returns_normalized_input_mean_rstd(
-    eps: float,
-    cudnn_enabled: int,
+    use_input_stats: bool,
+    momentum: Number,
```
We have seen cases where momentum is passed in as an int. The torch `Number` type consists of int, float and bool (which are 0 1 ints). So I thought to leverage it instead of creating our own number type or writing Union[int, float]

BowenBao replied:
I see, although I think it should got cast to float when annotated with "f" in `parse_args` though?

justinchuby replied:
It's possible then it was called by other symbolics and the arg was int. In that case (point) It'd be good to enforce float to ensure consistency (counter point) existing users may call with int and get an error

BowenBao replied:
Hmm looks like the extra layer of decorator messes up the argument types here. This is interesting. It's an even broader problem for those "parsed" arguments, since at caller they are all `Value`. I guess we could type annotate this guy here https://github.com/pytorch/pytorch/blob/a423c966a780a1fdac6a29c6d2be2a0709de2cd5/torch/onnx/symbolic_helper.py#L283 Still feels a bit sketchy.

justinchuby replied:
I am going to fix the caller (Quantized instance norm). But I think it makes sense to accept a (real) Number type for anything that is supposed to be floats. With that we reduce the risk of throwing errors because of callers we don't control. I can't think of anything harmful. WDYT?

justinchuby replied:
Even `True / 2 = 0.5`. I think we are safe

BowenBao replied:
That's ok. I'm a bit more concerned about the `parse_args` decorator altering argument type without being (or is it possible to) properly annotated.

justinchuby replied:
Resolved offline

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

TODO: profile performance

[ghstack-poisoned]

BowenBao commented:
> > LGTM. I noticed that some new annotations were added, while others were not. Is that for passing certain check failures?
> 
> I may have missed some. Could you point them out?

nvm, I was referring to type annotations, and I think I mistook a few updates as new annotations.

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

TODO: profile performance

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

TODO: profile performance

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

TODO: profile performance

[ghstack-poisoned]

justinchuby commented:
Profiler:

## type checking on

```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 02:17:56  Samples:  427
 /_//_/// /_\ / //_// / //_'/ //     Duration: 2.291     CPU time: 5.753
/   _/                      v4.3.0

Program: /home/justinchu/dev/pytorch/alexnet.py

2.291 <module>  bug.py:1
└─ 2.291 export  <@beartype(torch.onnx.utils.export) at 0x7fa306f42430>:1
   └─ 2.291 export  torch/onnx/utils.py:191
      └─ 2.289 _export  torch/onnx/utils.py:1370
         ├─ 1.282 _model_to_graph  <@beartype(torch.onnx.utils._model_to_graph) at 0x7fa306f5b280>:1
         │  └─ 1.282 _model_to_graph  torch/onnx/utils.py:1030
         │     ├─ 0.794 _optimize_graph  <@beartype(torch.onnx.utils._optimize_graph) at 0x7fa306f42700>:1
         │     │  └─ 0.794 _optimize_graph  torch/onnx/utils.py:543
         │     │     ├─ 0.387 _run_symbolic_function  torch/onnx/utils.py:1748
         │     │     │  ├─ 0.200 register_version  <@beartype(torch.onnx.symbolic_registry.register_version) at 0x7fa30703e9d0>:1
         │     │     │  │  └─ 0.200 register_version  torch/onnx/symbolic_registry.py:52
         │     │     │  │     └─ 0.200 register_ops_in_version  <@beartype(torch.onnx.symbolic_registry.register_ops_in_version) at 0x7fa306f35310>:1
         │     │     │  │        └─ 0.200 register_ops_in_version  torch/onnx/symbolic_registry.py:67
         │     │     │  │           └─ 0.200 register_ops_helper  <@beartype(torch.onnx.symbolic_registry.register_ops_helper) at 0x7fa306f351f0>:1
         │     │     │  │              └─ 0.200 register_ops_helper  torch/onnx/symbolic_registry.py:60
         │     │     │  │                 ├─ 0.146 get_ops_in_version  <@beartype(torch.onnx.symbolic_registry.get_ops_in_version) at 0x7fa306f354c0>:1
         │     │     │  │                 │  └─ 0.146 get_ops_in_version  torch/onnx/symbolic_registry.py:98
         │     │     │  │                 │     ├─ 0.088 getmembers  inspect.py:326
         │     │     │  │                 │     │     [20 frames hidden]  inspect, <built-in>
         │     │     │  │                 │     └─ 0.041 [self]  
         │     │     │  │                 └─ 0.047 is_registered_op  <@beartype(torch.onnx.symbolic_registry.is_registered_op) at 0x7fa306f35820>:1
         │     │     │  │                       [6 frames hidden]  <@beartype(torch.onnx.symbolic_regist...
         │     │     │  └─ 0.142 Constant  <@beartype(torch.onnx.symbolic_opset9.Prim.Constant) at 0x7fa306e8c310>:1
         │     │     │     └─ 0.142 Constant  torch/onnx/symbolic_opset9.py:6104
         │     │     │        └─ 0.133 _graph_op  <@beartype(torch.onnx._patch_torch._graph_op) at 0x7fa307246940>:1
         │     │     │              [2 frames hidden]  <@beartype(torch.onnx._patch_torch
         │     │     │                 0.132 _graph_op  torch/onnx/_patch_torch.py:19
         │     │     │                 └─ 0.117 PyCapsule._jit_pass_onnx_node_shape_type_inference  <built-in>:0
         │     │     │                       [2 frames hidden]  <built-in>
         │     │     ├─ 0.315 PyCapsule._jit_pass_onnx_graph_shape_type_inference  <built-in>:0
         │     │     │     [2 frames hidden]  <built-in>
         │     │     └─ 0.051 [self]  
         │     ├─ 0.306 PyCapsule._jit_pass_onnx_graph_shape_type_inference  <built-in>:0
         │     │     [2 frames hidden]  <built-in>
         │     └─ 0.178 _create_jit_graph  torch/onnx/utils.py:906
         │        └─ 0.177 _trace_and_get_graph_from_model  torch/onnx/utils.py:847
         │           └─ 0.176 _get_trace_graph  torch/jit/_trace.py:1138
         │              └─ 0.176 _call_impl  torch/nn/modules/module.py:1184
         │                 └─ 0.176 forward  torch/jit/_trace.py:94
         │                    └─ 0.160 wrapper  torch/jit/_trace.py:104
         │                       ├─ 0.109 _call_impl  torch/nn/modules/module.py:1184
         │                       │  └─ 0.108 _slow_forward  torch/nn/modules/module.py:1164
         │                       │     └─ 0.108 forward  torchvision/models/alexnet.py:47
         │                       │           [3 frames hidden]  torchvision, <built-in>
         │                       │              0.107 _call_impl  torch/nn/modules/module.py:1184
         │                       │              └─ 0.107 _slow_forward  torch/nn/modules/module.py:1164
         │                       │                 └─ 0.102 forward  torch/nn/modules/container.py:202
         │                       │                    └─ 0.102 _call_impl  torch/nn/modules/module.py:1184
         │                       │                       └─ 0.102 _slow_forward  torch/nn/modules/module.py:1164
         │                       │                          └─ 0.057 forward  torch/nn/modules/conv.py:462
         │                       │                             └─ 0.057 _conv_forward  torch/nn/modules/conv.py:454
         │                       │                                └─ 0.056 _VariableFunctionsClass.conv2d  <built-in>:0
         │                       │                                      [2 frames hidden]  <built-in>
         │                       └─ 0.051 <genexpr>  torch/jit/_trace.py:114
         │                          └─ 0.043 Parameter.clone  <built-in>:0
         │                                [2 frames hidden]  <built-in>
         ├─ 0.703 [self]  
         ├─ 0.201 PyCapsule._check_onnx_proto  <built-in>:0
         │     [2 frames hidden]  <built-in>
         └─ 0.101 BytesIO.write  <built-in>:0
               [2 frames hidden]  <built-in>
```

## Type checking off

```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 02:28:10  Samples:  376
 /_//_/// /_\ / //_// / //_'/ //     Duration: 2.279     CPU time: 5.658
/   _/                      v4.3.0

Program: /home/justinchu/dev/pytorch/alex.py

2.278 <module>  bug.py:1
└─ 2.278 export  torch/onnx/utils.py:191
   └─ 2.276 _export  torch/onnx/utils.py:1370
      ├─ 1.249 _model_to_graph  torch/onnx/utils.py:1030
      │  ├─ 0.755 _optimize_graph  torch/onnx/utils.py:543
      │  │  ├─ 0.367 _run_symbolic_function  torch/onnx/utils.py:1748
      │  │  │  ├─ 0.172 register_version  torch/onnx/symbolic_registry.py:52
      │  │  │  │  └─ 0.172 register_ops_in_version  torch/onnx/symbolic_registry.py:67
      │  │  │  │     └─ 0.172 register_ops_helper  torch/onnx/symbolic_registry.py:60
      │  │  │  │        └─ 0.156 get_ops_in_version  torch/onnx/symbolic_registry.py:98
      │  │  │  │           ├─ 0.080 getmembers  inspect.py:326
      │  │  │  │           │     [18 frames hidden]  inspect, <built-in>
      │  │  │  │           └─ 0.051 [self]  
      │  │  │  └─ 0.148 Constant  torch/onnx/symbolic_opset9.py:6104
      │  │  │     └─ 0.146 _graph_op  torch/onnx/_patch_torch.py:19
      │  │  │        └─ 0.143 PyCapsule._jit_pass_onnx_node_shape_type_inference  <built-in>:0
      │  │  │              [2 frames hidden]  <built-in>
      │  │  ├─ 0.319 PyCapsule._jit_pass_onnx_graph_shape_type_inference  <built-in>:0
      │  │  │     [2 frames hidden]  <built-in>
      │  │  └─ 0.028 [self]  
      │  ├─ 0.312 PyCapsule._jit_pass_onnx_graph_shape_type_inference  <built-in>:0
      │  │     [2 frames hidden]  <built-in>
      │  └─ 0.177 _create_jit_graph  torch/onnx/utils.py:906
      │     └─ 0.176 _trace_and_get_graph_from_model  torch/onnx/utils.py:847
      │        └─ 0.176 _get_trace_graph  torch/jit/_trace.py:1138
      │           └─ 0.175 _call_impl  torch/nn/modules/module.py:1184
      │              └─ 0.175 forward  torch/jit/_trace.py:94
      │                 └─ 0.158 wrapper  torch/jit/_trace.py:104
      │                    ├─ 0.110 _call_impl  torch/nn/modules/module.py:1184
      │                    │  └─ 0.110 _slow_forward  torch/nn/modules/module.py:1164
      │                    │     └─ 0.109 forward  torchvision/models/alexnet.py:47
      │                    │           [3 frames hidden]  torchvision, <built-in>
      │                    │              0.108 _call_impl  torch/nn/modules/module.py:1184
      │                    │              └─ 0.108 _slow_forward  torch/nn/modules/module.py:1164
      │                    │                 └─ 0.103 forward  torch/nn/modules/container.py:202
      │                    │                    └─ 0.103 _call_impl  torch/nn/modules/module.py:1184
      │                    │                       └─ 0.103 _slow_forward  torch/nn/modules/module.py:1164
      │                    │                          └─ 0.060 forward  torch/nn/modules/conv.py:462
      │                    │                             └─ 0.060 _conv_forward  torch/nn/modules/conv.py:454
      │                    │                                └─ 0.060 _VariableFunctionsClass.conv2d  <built-in>:0
      │                    │                                      [2 frames hidden]  <built-in>
      │                    └─ 0.048 <genexpr>  torch/jit/_trace.py:114
      │                       └─ 0.044 Parameter.clone  <built-in>:0
      │                             [2 frames hidden]  <built-in>
      ├─ 0.717 [self]  
      ├─ 0.205 PyCapsule._check_onnx_proto  <built-in>:0
      │     [2 frames hidden]  <built-in>
      └─ 0.104 BytesIO.write  <built-in>:0
            [2 frames hidden]  <built-in>
```

justinchuby commented:
@pytorchbot merge -g

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1695,14 +1744,15 @@ def _get_aten_op_overload_name(n: _C.Node) -> str:
     return _C.parse_schema(schema).overload_name
 
 
+# @_beartype.beartype
```
Forgot to turn this on #84170

justinchuby replied:
Done

weiwangmeta commented:
@pytorchbot merge -l

weiwangmeta commented:
@pytorchbot merge -f "trunk signal macos failure not related to this PR"

thiagocrepaldi reviewed (changes requested):
Overall looks promising. This is a huge change, with equally huge impact for future debugging. Let's take our time to make type checking as strict as possible. That will save lots of debugging time, which can be used for more generic improvements/feature support

A few general comments are below:

There are decorated functions/methods without annotation, which seems odd. 
Either 1) we annotate and enforce type checking , or 2) we don't annotate (adding a comment as to why it is not needed for future tracing and easy identification of missing spots) and don't add bear decorator. Otherwise, the decorator is dead code.

There are decorated methods/functions with relaxed types due to existing non-ideal implementation (python Tensors being passed as arguments to functions that should only deal with JIT types.
The goal of enforce type checking is capturing unexpected types which leads to unexpected behavior. I recommend taking this opportunity to identify current misuses and fix them. That will harden the checks, making it more valuable to us. Identifying relaxed annotations after the code is merged is harder because the reason for relaxing is not documented in a comment, and other engineers that didn't implement the type check are not aware of the relaxing. Even for the author, it is hard to remember about corner cases after a couple days working in something else.

There are methods/functions that have proper input annotation, but not return types. Returning the wrong type can lead to unexpected behavior as well, so ideally we should take care of that too.

justinchuby commented on torch/onnx/_patch_torch.py:
```diff
@@ -138,6 +144,7 @@ def _new_node(
 
 
+@_beartype.beartype
 def _is_onnx_list(value):
```
They don't do anything. I just added it for everything so when more annotations are added they can start functioning right away.

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -182,6 +198,7 @@ def _compare_ort_pytorch_outputs(
             raise
 
 
+@_beartype.beartype
```
Same comments as before. Added for everything so that future authors are not confused about when some functions don't have the decorators and reviewers don't have to look for them every time. When annotations are added they will start to function automatically

justinchuby commented:
This PR was merged and reverted I suppose to revert a Previous PR it depends on. I will rebase and add the missing changes

justinchuby commented on torch/onnx/_patch_torch.py:
```diff
@@ -146,19 +153,22 @@ def _is_onnx_list(value):
 
 
+@_beartype.beartype
 def _is_caffe2_aten_fallback():
```
Done. If not annotated, the implied type is `Any`

justinchuby commented on torch/onnx/symbolic_helper.py:
```diff
@@ -94,6 +97,7 @@
 
 
+@_beartype.beartype
 def _parse_arg(
```
_parse_arg can return a lot of different types. Annotating the return type will make mypy mad at various callers because the return value can be any one of those specify. I suggest dealing with it in a separate change

justinchuby commented on torch/onnx/symbolic_helper.py:
```diff
@@ -430,13 +448,15 @@ def wrapper(g, *args, **kwargs):
 
 
+@_beartype.beartype
 def _if_scalar_type_as(g: _C.Graph, self, tensor):
```
self and tensor can be complicated. Will handle it in the next pr

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

Profile:

export `torchvision.models.alexnet(pretrained=True)`

```
with runtime type checking: 21.314 / 10 passes
without runtime type checking: 20.797 / 10 passes

+ 2.48%
```

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

Profile:

export `torchvision.models.alexnet(pretrained=True)`

```
with runtime type checking: 21.314 / 10 passes
without runtime type checking: 20.797 / 10 passes

+ 2.48%
```

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

Profile:

export `torchvision.models.alexnet(pretrained=True)`

```
with runtime type checking: 21.314 / 10 passes
without runtime type checking: 20.797 / 10 passes

+ 2.48%
```

[ghstack-poisoned]

justinchuby commented on torch/onnx/symbolic_helper.py:
```diff
@@ -430,13 +448,15 @@ def wrapper(g, *args, **kwargs):
 
-def _scalar(x: torch.Tensor):
+@_beartype.beartype
+def _scalar(x: Any) -> Optional[Number]:
```
Acknowledged: will follow up

justinchuby commented on torch/onnx/symbolic_helper.py:
```diff
@@ -163,33 +167,41 @@ def _parse_arg(
 
 
+@_beartype.beartype
 def _get_const(value, desc, arg_name):
```
Acknowledged: will follow up

justinchuby commented on torch/onnx/_patch_torch.py:
```diff
@@ -115,6 +120,7 @@ def _block_op(b: _C.Block, opname: str, *args, **kwargs):
 
+@_beartype.beartype
 def _new_node(
     g: _C.Graph, namespace: str, op: str, outputs: int, *args, **kwargs
```
Done

justinchuby commented on torch/onnx/_patch_torch.py:
```diff
@@ -96,6 +100,7 @@ def _aten_op(g: _C.Graph, operator: str, *args, overload_name: str = "", **kwarg
 
 
+@_beartype.beartype
 def _block_op(b: _C.Block, opname: str, *args, **kwargs):
```
Done

justinchuby commented:
@thiagocrepaldi PTAL cc @BowenBao

justinchuby commented:
The relaxed types are there to match the reality. We can optimize the code after checks are turned on. The follow ups are also going to be more focused and easier to review

thiagocrepaldi commented:
> LGTM. I noticed that some new annotations were added, while others were not. Is that for passing certain check failures?

same concern here. I'd rather seeing a feature complete on a single PR than spreading towards multiple ones. If it needs to be reverted or cherry picked in the feature, due to a bug or limitation, it is easy to do in a single pr. It is also easier to read a single PR and learn about the introduced feature

thiagocrepaldi reviewed (approved):
Would rather seeing PRs being code-complete before merge instead of having many follow ups PRs.

That makes reverting unrelated PRs very hard, especially using pytorchbot. (in order to disable beartype, many PRs related and unrelated would have to be reverted (the ones in between all beartype PRs)

It is also harder to learn about the feature because there are too many PRs for the same task

replacing `_is_value` by `isinstance(tensor, _C.Value)` defeats the purposa of having `is_value()`entirely. At least a #TODO or an issue should be created to track and properly fix it.

justinchuby commented:
> Would rather seeing PRs being code-complete before merge instead of having many follow ups PRs.
> 
> That makes reverting unrelated PRs very hard, especially using pytorchbot. (in order to disable beartype, many PRs related and unrelated would have to be reverted (the ones in between all beartype PRs)

I like the idea of making revert easier. In this particular case, the bigger the PR becomes and the more places it touches, the harder it can be reverted. Future PRs would more likely need to be reverted because they will build on top of this PR once merged, based on the extend of files it would touch.

IMO this PR stands on its own in the sense that it _enables_ typing checking without substantial changes to the actual code. If we lump the logic changes as well, it will have a _higher_ chance of breaking things and being reverted.

Separating changes to bite sized self-contained PRs often means a smaller chance of breaking things, so we don't need to revert _everything_. When we start annotating more functions and refactor, the effort can be distributed and verified individually.

I suggest we keep PRs small and focused. https://google.github.io/eng-practices/review/developer/small-cls.html#cant

> It is also harder to learn about the feature because there are too many PRs for the same task

PR stacks have been there for this. Depending on what we need, we can create some type of documentation for the feature and do a good job of linking PRs together.

> replacing `_is_value` by `isinstance(tensor, _C.Value)` defeats the purposa of having `is_value()`entirely. At least a #TODO or an issue should be created to track and properly fix it.

Will do. thanks!

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

Profile:

export `torchvision.models.alexnet(pretrained=True)`

```
with runtime type checking: 21.314 / 10 passes
without runtime type checking: 20.797 / 10 passes

+ 2.48%
```

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

Profile:

export `torchvision.models.alexnet(pretrained=True)`

```
with runtime type checking: 21.314 / 10 passes
without runtime type checking: 20.797 / 10 passes

+ 2.48%
```

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Fix type annotations and enable type checking for all apis"


Enable runtime type checking for all torch.onnx public apis, symbolic functions and most helpers (minus two that does not have a checkable type: `_.JitType` does not exist) by adding the beartype decorator. Fix type annotations to makes unit tests green.

Profile:

export `torchvision.models.alexnet(pretrained=True)`

```
with runtime type checking: 21.314 / 10 passes
without runtime type checking: 20.797 / 10 passes

+ 2.48%
```

[ghstack-poisoned]

justinchuby commented:
@pytorchbot merge -g
